File: Path Sum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode *root, int sum) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        return checkPathSum(root, sum, false);
        
    }
    
    
    bool checkPathSum(TreeNode *root, int sum, bool isLeaf)
    {
        if(root==NULL)
        {
            if(isLeaf)
                return sum==0?true:false;
            else
                return false;
        }
        
        if(root->left==NULL&&root->right==NULL)
            isLeaf = true;
        
        int val = sum - root->val;
        
        bool left = checkPathSum(root->left, val, isLeaf);
        if(left)
            return true;
        else
        {
            bool right = checkPathSum(root->right, val, isLeaf);
            return right;
        }
        
    }
};

File: Path Sum.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(root == null)
        {
            return false;
        }
        sum -= root.val;
        if(root.left == null && root.right ==  null && sum ==0)
        {
            return true;
        }
        boolean leftBool = false;
        boolean rightBool = false;
        if(root.left != null)
        {
            leftBool = hasPathSum(root.left, sum);
        }
        
        if(!leftBool && root.right != null)
        {
            rightBool = hasPathSum(root.right, sum);
        }
        
        return leftBool||rightBool; 
    }
}